lib/tracker.js

```diff
@@ -1,35 +1,172 @@
+      info.instance = await this._createBee(publicKey, info, { keyEncoding: 'utf-8', valueEncoding: 'json' })
+    }
 
-    this.resources.push(info)
+    this.resources.set(id, info)
 
     return info
   }
 
+  async remove (key) {
+    const info = this.get(key)
+
+    if (info.seeders) await info.seeders.destroy()
+    await info.instance.close()
+
+    this.resources.delete(info.key)
+  }
+
+  has (key) {
+    const id = HypercoreId.normalize(key)
+    return this.resources.has(id)
+  }
+
+  get (key) {
+    const id = HypercoreId.normalize(key)
+    const info = this.resources.get(id)
+    if (!info) throw new Error('Resource does not exists')
+    return info
+  }
+
   getByType (type) {
-    return this.resources.filter(r => r.type === type)
+    const arr = []
+    for (const [, info] of this.resources) {
+      if (info.type === type) arr.push(info)
+    }
+    return arr
+  }
+
+  find (filter) {
+    for (const [, info] of this.resources) {
+      if (filter(info)) return info
+    }
+    return null
+  }
+
+  async _createCore (publicKey, info, opts = {}) {
+    const core = this.store.get(publicKey)
+    await core.ready()
+
+    this._suitCore(core, info, opts)
+
+    return core
+  }
+
+  async _createBee (publicKey, info, opts = {}) {
+    const core = this.store.get(publicKey)
+    const bee = new Hyperbee(core, { keyEncoding: opts.keyEncoding, valueEncoding: opts.valueEncoding })
+    await bee.ready()
+
+    this._suitCore(bee.core, info, opts)
+
+    return bee
   }
 
-  static speedometer (core, info) {
-    core.on('download', onspeed.bind(null, 'download', info))
-    core.on('upload', onspeed.bind(null, 'upload', info))
+  async _createDrive (publicKey, info, opts = {}) {
+    const drive = new Hyperdrive(this.store, publicKey)
+    await drive.ready()
+
+    this._suitCore(drive.core, info, opts)
+
+    const onblobs = (blobs) => this._suitCore(blobs.core, info, opts)
+    if (drive.blobs) onblobs(drive.blobs)
+    else drive.once('blobs', onblobs)
+
+    if (opts.seeders) {
+      info.seeders = await this._createSeeders(publicKey)
+    }
+
+    return drive
   }
+
+  async _createSeeders (publicKey) {
+    const id = HypercoreId.encode(publicKey)
+    const keyPair = await this.store.createKeyPair('simple-seeder-swarm@' + id)
+
+    const sw = new Seeders(publicKey, { dht: this.swarm.dht, keyPair })
+    sw.join()
+    sw.on('connection', (socket) => this.store.replicate(socket))
+
+    /* sw.on('update', function (record) {
+      // TODO: better tracking here
+      // tracking.notifs[id] = record
+    }) */
+
+    // const unregister = goodbye(() => sw.destroy())
+    // sw.on('close', () => unregister())
+
+    return sw
+  }
+
+  _suitCore (core, info, opts) {
+    followSpeed(core, info)
+    if (opts.announce !== false) this._swarmCore(core)
+    core.download()
+  }
+
+  _swarmCore (core) {
+    const done = core.findingPeers()
+    this.swarm.join(core.discoveryKey, { client: true, server: !this.backup })
```

Comment:
on need to store the discovery session returned by join and destroy it on if the resource is removed